doc/aaa/TACACS+ Design.md

```diff
@@ -0,0 +1,537 @@
+
+    // accounting with TACACS+ server
+    config aaa accounting tacacs
+
+    // accounting with local syslog
+    config aaa accounting local
+```
+
+ - Following command will disable accounting
+```
+    config aaa accounting disable
+```
+# 5 Error handling
+ - Bash plugin for authorization will return error code [Bash](#bash). and patched Bash will:
+   - Output error log to syslog.
+   - Output error message to stderr.
+ - [audisp-tacplus](#audisp-tacplus) will return errors as per [Auditd](#auditd)  respectively.
+
+# 6 Serviceability and Debug
+ - The Bash plugin and [audisp-tacplus](#audisp-tacplus) can be debugged by enabling the debug
+field of the AAA|authentication key. (Please see ConfigDB AAA Table
+Schema in [TACACS+ Authentication](#TACPLUS-Authentication)). 
+
+# 7 Unit Test
+## 7.1 Unit test for source code
+ - All patch code in Bash and Bash plugin should have 100% code coverage.
+ - Bash plugin test, all TACACS+ server not reachable test:
+```
+    Verify TACACS+ authorization failed.
+```
+
+ - Bash plugin test, partial TACACS+ server accessible, and user command config as allowed on all server.
+```
+    Verify TACACS+ authorization passed.
+```
+
+ - Bash plugin test, partial TACACS+ server accessible, and user command config as reject on all server.
+```
+    Verify TACACS+ authorization rejected.
+```
+
+ - Bash plugin test, partial TACACS+ server accessible, and user command config as reject on accessible server, and allow on not accessible server.
+```
+    Verify TACACS+ authorization rejected.
+```
+
+ - Bash plugin test, partial TACACS+ server accessible, and user command config as allow on accessible server, and reject on not accessible server.
+```
+    Verify TACACS+ authorization passed.
+```
+
+ - [audisp-tacplus](#audisp-tacplus) test, all TACACS+ server accessible.
+```
+    Verify TACACS+ accounting succeeded.
+```
+
+ - [audisp-tacplus](#audisp-tacplus) test, all TACACS+ server not accessible.
+```
+    Verify plugin return correct error code.
+```
+
+ - [audisp-tacplus](#audisp-tacplus) test, partial TACACS+ server accessible.
+```
+    Verify TACACS+ accounting succeeded.
+```
+
+ - [audisp-tacplus](#audisp-tacplus) test, user command longer than 240 bytes.
+```
+    Verify TACACS+ accounting succeeded.
+    Verify only 240 bytes of user command send to TACACS+ server side.
+```
+
+ - [audisp-tacplus](#audisp-tacplus) test, user command+parameter longer than 240 bytes.
+```
+    Verify TACACS+ accounting succeeded.
+    Verify only 240 bytes of user command+parameter send to TACACS+ server side.
+```
+
+## 7.2 End to end test
+
+- config AAA authorization with TACACS+ only:
+```
+    Verify TACACS+ user run command in server side whitelist:
+        If command have local permission, user can run command.
+        If command not have local permission, user can't run command.
+    Verify TACACS+ user can't run command not in server side whitelist.
+```
+
+- config AAA authorization with TACACS+ only:
```

Comment:
>config AAA authorization with TACACS+ only [](http://example.com/codeflow?start=2&length=42)

This a very practical work flow. Normally use will continue to login with local account, and run command as local authorization, and even tacacs server recover, it does not impact local user.

Reply:
config AAA authorization with TACACS+ and local, but server not accessible:

Add this step: when remote tacacs accessible, local user still can run command. doesn't impact local user authorization.

Check if we have PR cover this: when tacacs accessible, local can't login.

Reply:
Discuss offline, and you will add some step in another test case below.

Reply:
Fixed.